rfcs/20220309-feature_flags.md

```diff
@@ -0,0 +1,229 @@
+
+### States
+
+Each flag can have 3 states.
+
+- **enabled** (`on`)
+
+  There is no fencing, everyone can access the feature, even if they don't like or want it.
+
+- **switchable** (`any`)
+
+  The feature is fenced by default but a user can opt-in. Preferences may be recorded.
+
+- **disabled** (`off`)
+
+  The feature is fenced and inaccessible, at least not visible. What's ever hidden in open-source?
+
+### Sources
+
+Flags must first be defined somewhere from where they can be loaded into the application.
+
+#### 1. Environment variables
+
+##### Pros
+
+1. Effortless to set up.
+1. Universal way to configure applications.
+1. Low maintenance as every feature can be one env var.
+1. Fairly easy to test considering env vars can be set per-session, per-command, per-container and per-machine (as needed).
+1. Managing a couple of `.env` files is not hard.
+1. Server-side envs can help tree-shake the off features in prod.
+
+##### Cons
+
+1. Will need a deployment to change the environment variables.
+1. Can get complicated if we need to build a lot of scaffolding around it.
+1. Managing a lot of `.env` files is not easy.
+1. Flags will be primitive and will not allow complex data.
+1. Nested flags and flags with metadata will need hacks.
+
+##### Considerations
+
+Feature flag env vars can be prefixed with a string like `FF_` to separate them from the other env vars.
+
+The limitation for the data to be a string means any extra metadata must be JSON-stringified and then parsed in the code. Only primitive flag states can be handled conveniently.
+
+Nesting flags will use hacks like artificial separators e.g. `FF_PARENT__CHILD`. Looks ugly and confusing.
+
+#### 2. Query params
+
+##### Pros
+
+1. Even more effortless to set up.
+1. We've done this before for the embedded mode, [_remember_](https://github.com/WordPress/openverse-frontend/pull/65)?
+1. Testing is easy because tests can append the right flag param to the URL.
+
+##### Cons
+
+1. Configuration will need to live in the codebase so change of state will need deployment.
+1. Exposes half-baked features to users who know the flag.
+1. Transient, so flags must be set on every URL or stored in cookies.
+1. Flags will be primitive and will not allow complex data. Even stringified JSON would be bad in URLs.
+
+##### Considerations
+
+We can clearly demarcate feature flag params with a prefix such as `ff_` (or `feature/` 🚳) to separate them from the actual query params and process them separately.
+
+Due to their nature being run-time rather than build-time, they are best used in pairing with another storage (as an override for the default) rather than on their own.
+
+For example, we can have a feature flag which is set to disabled via a config file. This prevents users from accessing this feature. However, a developer who still wants to use this feature can send the query param as true and access it.
```

Comment:
I don't think this matches the description of the flag states above. If a flag is `disabled` it's completely inaccessible in that environment. Only `switchable` flags can be manipulated. `enabled` flags are always enabled and cannot be manipulated either.

Reply:
You're right, I wanted the params to override the config instead of supplementing it. That was not valid and has been dropped in the last update.